src/box/box.cc

```diff
@@ -1805,6 +1805,214 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+	uint32_t id = ack->source;
+	uint64_t sync = data->syncs[id];
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->sync >= sync) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	try {
+		if (!bit_test(&data->collected_vclock_map, id)) {
+			data->syncs[id] = relay_next_sync(relay,
+							  data->deadline);
+		}
+	} catch (TimedOut *) {
+		data->timed_out = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** Find the minimal vclock which has all the data confirmed on a quorum. */
+static int
+box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
+{
+	/*
+	 * How many vclocks we should see to be sure that at least one of them
+	 * contains all data present on any real quorum.
+	 */
+	int vclock_count = MAX(1, replicaset.registered_count) -
+			   replication_synchro_quorum + 1;
+	/*
+	 * We should check the vclock on self plus vclock_count - 1 remote
+	 * instances.
+	 */
+	vclock_copy(confirmed_vclock, &replicaset.vclock);
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.syncs[replica->id] == 0) {
+			data.syncs[replica->id] =
+				relay_next_sync(replica->relay, deadline);
+		}
+	}
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon)
+			continue;
+		uint64_t sync = data.syncs[replica->id];
+		struct relay *relay = replica->relay;
+
+		/* Might be already accounted by on_ack trigger. */
+		if (!bit_test(&data.collected_vclock_map, replica->id) &&
+		    sync > 0 && relay_sync(relay) >= sync) {
+			vclock_max_ignore0(confirmed_vclock,
+					   relay_vclock(relay));
+			bit_set(&data.collected_vclock_map, replica->id);
+		}
+	}
+
+	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
```

Comment:
Yes.
Imagine, you have a cluster of 3 nodes, quorum is 2. Then the user sets quorum to 3. It means that linearizable read doesn't need to visit remote nodes anymore to complete. If existing linearizable read requests stopped waiting for vclocks because of quorum increase, it would make them see old data (because there still might be some requests that were on 2 other nodes but haven't yet reached this node).

And vice versa. When out of 3 nodes quorum was 2, and then became 1, we don't have to wait for all 3 nodes. Because at the start of linearizable read quorum was 2, so all the requests that were committed at the moment linearizable read started are definitely on 2 out of 3 replicas.

Reply:
Thanks for the explanation, sounds legit.